controllers/etcd_controller.go

```diff
@@ -576,64 +587,112 @@ func (r *EtcdReconciler) reconcileStatefulSet(cm *corev1.ConfigMap, svc *corev1.
 		for i := 1; i < len(filteredStatefulSets); i++ {
 			ss := filteredStatefulSets[i]
 			if err := r.Delete(context.TODO(), ss); err != nil {
-				logger.Error(err, "Error in deleting duplicate StatefulSet")
+				logger.Error(err, "error in deleting duplicate statefulset")
 				continue
 			}
 		}
 
 		// Return the updated statefulset
 		ss := &appsv1.StatefulSet{}
 		if err := r.Get(context.TODO(), types.NamespacedName{Name: filteredStatefulSets[0].Name, Namespace: filteredStatefulSets[0].Namespace}, ss); err != nil {
-			return nil, err
+			return nil, &reconcileResultRequeue, druidv1alpha1.LastOperationTypeReconcile, err
+		}
+
+		pvcList := &corev1.PersistentVolumeClaimList{}
+		if handleScaleDown {
+			// if volume(s) are already marked for deletion, check if they are deleted
+			if etcd.Status.LastOperation.Type == druidv1alpha1.LastOperationTypeVolumeDelete {
+				ctx, cancel := context.WithTimeout(context.TODO(), DefaultTimeout)
+				defer cancel()
+				arePVCsDeleted, err := r.arePVCsDeleted(ctx, ss)
+				if err != nil {
+					return nil, &reconcileResultRequeueAfter, druidv1alpha1.LastOperationTypeVolumeDelete, err
+				}
+				if !arePVCsDeleted {
+					return ss, &reconcileResultRequeueAfter, druidv1alpha1.LastOperationTypeVolumeDelete, nil
+				}
+
+				return ss, &reconcileResultRequeueNone, druidv1alpha1.LastOperationTypeReconcile, nil
+			}
+
+			ctx, cancel := context.WithTimeout(context.TODO(), DefaultTimeout)
+			defer cancel()
+			pvcList, err = r.prepareEtcdForScaleDown(ctx, etcd, svc)
+			if err != nil {
+				return nil, &reconcileResultRequeueAfter, druidv1alpha1.LastOperationTypeReconcile, err
+			}
 		}
 
 		// Statefulset is claimed by for this etcd. Just sync the specs
 		if ss, err = r.syncStatefulSetSpec(ss, cm, svc, etcd, values); err != nil {
-			return nil, err
+			return nil, &reconcileResultRequeue, druidv1alpha1.LastOperationTypeReconcile, err
 		}
 
 		// restart etcd pods in crashloop backoff
 		selector, err := metav1.LabelSelectorAsSelector(ss.Spec.Selector)
 		if err != nil {
 			logger.Error(err, "error converting statefulset selector to selector")
-			return nil, err
+			return nil, &reconcileResultRequeue, druidv1alpha1.LastOperationTypeReconcile, err
 		}
 		podList := &v1.PodList{}
 		if err := r.List(context.TODO(), podList, client.InNamespace(etcd.Namespace), client.MatchingLabelsSelector{Selector: selector}); err != nil {
-			return nil, err
+			return nil, &reconcileResultRequeue, druidv1alpha1.LastOperationTypeReconcile, err
 		}
 
-		for _, pod := range podList.Items {
+		errs := make([]error, len(podList.Items))
+		for i, pod := range podList.Items {
 			if utils.IsPodInCrashloopBackoff(pod.Status) {
 				if err := r.Delete(context.TODO(), &pod); err != nil {
 					logger.Error(err, fmt.Sprintf("error deleting etcd pod in crashloop: %s/%s", pod.Namespace, pod.Name))
-					return nil, err
+					errs[i] = err
 				}
 			}
 		}
+		for _, err := range errs {
```

Comment:
Why not use something [errors aggregate](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/errors/errors.go#L35)?